iowrapper: add kdbiowrapper.h & libuv io-wrapper

waht opened the pull request:
# Purpose

Includes the io-wrapper interface and io-wrapper for libuv. These are preparations for io handling of the upcoming notification plugins. Developer documentation (usage in *elektrified* applications) and adapters for DBus, ZeroMQ and Redis are not yet included. Without adapters the code cannot be tried yet.

**Building:**
Enable the libuv-io-wrapper by setting the new ``IOWRAPPERS`` CMake-variable to "libuv" - default is "", so no io-wrappers are built by default. The libuv-io-wrapper requires libuv >= 1.9.

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went -fine- (some markdown tests failed for me with and without the io-wrapper - I didn't investigate this further)
- [x] I added code comments, logging, and assertions

@markus2330 please review my pull request

markus2330 commented on cmake/ElektraCache.cmake:
```diff
@@ -185,6 +185,42 @@ set(TOOLS ${TOOLS} CACHE STRING ${TOOLS_DOC} FORCE)
+# List of IO-Wrappers to build
+#
+
+remember_for_removal(IOWRAPPERS TO_REMOVE_IOWRAPPERS)
```
This macro is a crude cmake hack and was (for good reasons) already replaced for plugins (bindings and tools unfortunately still use it). Would be better if we do not rely new stuff on it. Maybe you can simply have the iowrappers as plugins or bindings?

Alternatively, we should introduce a `LIBS` variable.

waht replied:
Good point! I will use BINDINGS like the Swig bindings do (e.g. io_uv, io_glib, ALL and IO shortcut)

markus2330 commented on src/include/kdbiowrapper.h:
```diff
@@ -0,0 +1,195 @@
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+#ifndef KDB_IORAPPER_H_
+#define KDB_IORAPPER_H_
```
(W)RAPPER?

waht replied:
Haha yes not Rapper 😎

markus2330 commented on src/include/kdbiowrapper.h:
```diff
@@ -0,0 +1,195 @@
+
+struct ElektraIoWrapper;
+
+typedef struct ElektraIoWrapper
```
In general very nice docu, but here explanation is missing.

Why do we need ElektraIoWrapper and its pointers to functions?

markus2330 commented on src/libs/iowrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+if (FINDEX GREATER -1)
+	message (STATUS "Include IO-Wrapper for libuv")
+	add_subdirectory (libuv)
+endif()
```
if not found, it should be excluded.

markus2330 commented on src/libs/iowrapper/libuv-wrapper.c:
```diff
@@ -0,0 +1,368 @@
+ */
+static int flagsToEvents (int flags)
+{
+	// fprintf(stderr, "flagsToEvents\n");
```
please use logger (and output flags?)

markus2330 commented on src/libs/iowrapper/libuv-wrapper.c:
```diff
@@ -0,0 +1,368 @@
+static void LibUvWrapper_HandleClosedCallback (uv_handle_t * handle)
+{
+	// fprintf(stderr, "LibUvWrapper_HandleClosedCallback\n");
+	WrapperLibuvData * p = (WrapperLibuvData *)handle->data;
```
Please assert handle to be not null.

markus2330 commented on src/libs/iowrapper/libuv-wrapper.c:
```diff
@@ -0,0 +1,368 @@
+ *
+ * @param handle any libuv handle type (e.g. poll, timeout, idle)
+ */
+static void LibUvWrapper_HandleClosedCallback (uv_handle_t * handle)
```
Please avoid `_` in function names: only camel case.

markus2330 commented on src/libs/iowrapper/libuv-wrapper.c:
```diff
@@ -0,0 +1,368 @@
+
+ElektraIoWrapper * LibUvWrapper_Init (uv_loop_t * loop)
+{
+	ElektraIoWrapper * wrapper = (ElektraIoWrapper *)malloc (sizeof (*wrapper));
```
Please use `elektraMalloc`.

markus2330 commented on src/libs/iowrapper/libuv-wrapper.c:
```diff
@@ -0,0 +1,368 @@
+	return 0;
+}
+
+ElektraIoWrapper * LibUvWrapper_Init (uv_loop_t * loop)
```
Better `New` instead of `Init`: you are allocating it, not only initializing it.

markus2330 commented on src/libs/iowrapper/libuv/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+
+	target_link_libraries (${IOWRAPPER_LIBUV_LIBRARY} PUBLIC ${LIBUV_LDFLAGS})
+
+  install (
```
indentation?

markus2330 commented on src/libs/iowrapper/libuv/libiowrapper_libuv.pc.in:
```diff
@@ -0,0 +1,14 @@
+tool_execdir=${prefix}/@TARGET_TOOL_EXEC_FOLDER@
+templatedir=${prefix}/@TARGET_TEMPLATE_FOLDER@
+
+Name: libiowrapper_libuv
```
Too general name? Maybe libelektra-io-uv?

markus2330 commented on src/libs/iowrapper/libuv/libuv-wrapper.c:
```diff
@@ -0,0 +1,360 @@
+{
+	WrapperLibuvData * privateData = (WrapperLibuvData *)handle->data;
+
+	free (privateData);
```
Please use elektraFree.

waht commented:
@markus2330 I moved the wrapper to bindings, please review

waht commented:
@markus2330 I've added the tutorial and test-suite. Please review :)

Do you have any idea why the Mac OSX build fails? It seems like it does not link against Elektra (altough I'm using `target_link_elektra`).

Also how do I add a library to build.elektra.org?

Thank you!

markus2330 commented:
> Do you have any idea why the Mac OSX build fails? It seems like it does not link against Elektra (altough I'm using target_link_elektra).

MacOSX is more strict in that plugins need to link against all libraries as needed at run-time. (glibc is a bit more sloppy: if the libs are loaded within the process, it works: even if the plugin is not linked against some of its deps). Maybe you need to link against some further libuv lib?

> Also how do I add a library to build.elektra.org?

Simply request it in #160. I installed libuv1-dev on 4 agents if it is missing somewhere (or you wanted something else to be installed) please ping @BernhardDenner and me in #160.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,235 @@
+
+**The features described in this document are not implemented yet or experimental. This document contains a preview of how developers would use the notification feature**
+
+This document explains how notifications are implemented in Elektra and how they can be used by application developers.
```
Could you please format the text with `fmt` or at least make a linebreak for every sentence? Doing so makes the review easier to understand (which comment refers to which text).

waht commented:
Great, thank you! The io_uv binding now builds on Linux and OSX - on Windows it is disabled.

I relaxed the timing requirements: Now warnings are printed when intervals deviate more than 5ms and tests fail for 500ms+ - i.e. when the binding is broken. The warnings are inteded for binding devs that execute tests locally. They also give more information about expected and measured values.

If everything is fine, all I need to do is some cleanup (e.g. remove io_uv from default bindings, remove cmake debug statements, squash commits).

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+ Developers do not interact with
+those plugins directly.
+ The underlying transports are transparent
+to them.
```
to?

waht replied:
I think the preposition is right. I found [a discussion on dict.leo.org](https://dict.leo.org/forum/viewWrongentry.php?idThread=1233522&idForum=3&lang=de&lp=ende) and [this page](https://lingohelp.me/preposition-after-adjective/transparent-about-to-in-with-for/).

markus2330 commented on doc/tutorials/notifications.md:
````diff
@@ -0,0 +1,327 @@
+to use a transport.
+ 
+```sh
+kdb mount file.ini system/example ini dbussend dbusreceive
````
dbussend currently called dbus (incompatible change?)

waht replied:
Not necessarily. I was thinking about adding either `dbussend` or `dbus` (and in that case changing the plugin's name) to the `infos/provides` field of the plugin contract.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+ This is suitable for programs where
+the Key's value is simply displayed or used repeatedly (e.g. by a timer
+or in a loop).
+ When the program's initialization depends on the value
```
Structure of text: First explain how variable registering works, then explain where its limits are and how to use callbacks.

markus2330 commented on doc/tutorials/notifications.md:
````diff
@@ -0,0 +1,327 @@
+int someKeyValue;
+
+elektraNotificationRegisterInt (repo, &someKeyValue, someKey);
+```
````
add loop or similar to show what it does. Maybe as separate example program?

waht replied:
Check. Once internalnotification plugin and notification wrapper are ready I will add a example program.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+
+
+/**
+ * Add idle task to IO-Binding
```
What is an idle task here? Why do we need to implement it?

waht replied:
This is just an example of a function for adding any [timer|idle|fd]-operation - But it's better to add it to the description :)

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+
+/**
+ * Create and initialize a new IO binding.
+ * @param  loop Loop to use for IO operations
```
Which param?

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+		return NULL;
+	}
+	// Set function pointers
+	binding->addFd = myElektraIoBindingAddFd;
```
Where do we get this from?

Example seems to be too big for this tutorial. Maybe add something similar to src/plugins/doc?

waht replied:
Yes the the full code would be too big. Having some kind of "doc"-binding is a good idea.

markus2330 commented on src/bindings/io/test/test_fd.c:
```diff
@@ -0,0 +1,416 @@
+	testStop ();
+}
+
+static void testFdShouldUpdateFlags (ElektraIoTestSuiteCreateBinding createBinding, ElektraIoTestSuiteStart start,
```
Can you add a bit of explanations for the test cases? How is the pipe used?

waht replied:
Ok - maybe a primer before the test cases? `pipe` is used to create simple (and POSIX compliant) file descriptors for the tests. Plus it's easy to use them to test the flags (e.g. wait for readability)

markus2330 commented on src/bindings/io/test/test_utils.c:
```diff
@@ -0,0 +1,107 @@
+/**
+ * @file
+ *
+ * @brief Tests for haskelltemplate plugin
```
?

waht replied:
Ops...must have accidentially copied that along with the license header.

markus2330 commented on src/bindings/io/uv/README.md:
````diff
@@ -0,0 +1 @@
	uv_loop_close (loop);
	free (loop);
}
```
````
Line breaks broken? dos2unix?

waht replied:
Missed that when commiting from Windows.

markus2330 commented on src/include/kdbio.h:
```diff
@@ -0,0 +1,241 @@
+/**
+ * @file
+ *
+ * @brief
```
Purpose of this header file? Is this public?

waht replied:
Yes. It contains the ElektraIoInterface which is used by bindings, plugins and applications. I'll add a description.

markus2330 commented on src/include/kdbio.h:
```diff
@@ -0,0 +1,241 @@
+/**
+ * Information about a file descriptor watched by IO-Binding
+ */
+typedef struct ElektraIoFdOperation
```
If it is public we should avoid exposing structs because it would prohibit any extensions.

waht replied:
I see. This would require some kind of accessor (getter and setter) function for every struct member, right?
I inteded the structs to be used directly, e.g (from src/bindings/io/uv/test/test_fd.c):
```C
ElektraIoFdOperation * fdOp = elektraIoTestSuiteUtilNewFdOperation ();  // malloc
fdOp->fd = fds[1];
fdOp->flags = ELEKTRA_IO_WRITABLE;
fdOp->enabled = 1;
fdOp->callback = testFdShouldSignalWritableProbe;

ElektraIoInterface * binding = createBinding (); // malloc + set pointers
binding->addFd (binding, fdOp);
```
But this means the code would transform into:
```C
// with a combined creator function (malloc + set struct members)
ElektraIoFdOperation * fdOp = elektraIoNewFdOperation (
  fds[1],  // int fd
  ELEKTRA_IO_WRITABLE,  // int flags
  1,  // int enabled
  NULL,  // void * privateData
  testFdShouldSignalWritableProbe  // callback
);

ElektraIoInterface * binding = createBinding ();
// when the ElektraIoInterface struct is not public, 
// this would also require functions to call binding functions
// instead of binding->addFd (binding, fdOp);
elektraIoBindingAddFd (binding, fdOp); 
```

I'm just asking because I'm not sure if that's what you meant; and it will take some work :)

I have to admit `elektraIoNewFdOperation()` is quite nice, because it will reduce possibilities for errors (e.g. forgot to fill a struct member).

markus2330 replied:
You got the idea. Get/set for every member is not always necessary, maybe your proposed constructor elektraIoNewFdOperation is enough. (Depending on if there are use cases which requires changes of already constructed FdOperation.) If getter/setter are required, maybe a single getter/setter is enough: one parameter could be used to tell which part should be changed, similar to the sysconf (3) call. In case of doubt (if something is needed), keep the API minimal.

waht commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+implemented yet or experimental. This document contains a preview of
+how developers would use the notification feature**
+ 
+ This document
```
Seems like it was introduced by the version of `fmt` I used. I'll remove those spaces.

markus2330 replied:
There were several strange line-breaking decisions, which version of `fmt` do you use? Seems I should not recommend the tool.

waht replied:
Maybe it was my version bundled with Git Bash on Windows or the flags I tried (haven't used it before).

waht commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,327 @@
+ 
+ Elektra's notification feature consists of several
+components.
+ While notification features are implemented by plugins,
```
Both sending & receiving - To clarify I changed it to "While sending and receiving notifications is implemented [...]"

waht commented on cmake/ElektraCache.cmake:
```diff
@@ -103,6 +103,14 @@ if (FINDEX GREATER -1 OR BINDINGS MATCHES "ALL")
 	set (BINDINGS_FORCE FORCE)
 endif()
 
+list (FIND BINDINGS "IO" FINDEX)
```
Ok. I'll look into that.

waht replied:
Alright. Seems like I'll have to update all bindings then. There is too much interference when I try to only change the behavior for the I/O bindings. I'll commit my changes and then change this.

Would it be nice to have build categories like plugins do (e.g. ALL;-EXPERIMENTAL)?

markus2330 replied:
> I'll commit my changes and then change this.

Thank you, such infrastructure work is really important.

> Would it be nice to have build categories like plugins do (e.g. ALL;-EXPERIMENTAL)?

Yes, categories would be important, especially categories like IO, SWIG, etc. to group bindings based on same technologies.

Ideally there would be reuse with the add_plugin logic so that we can easily extend to add_tool without duplicating CMake code three times.

waht replied:
Ok. I will try to refactor `plugin_check_if_included` to be more generic. I think it would be nice to keep the change as minimal as possible. For example:
```cmake
list (FIND BINDINGS "io_uv" FINDEX)
if (FINDEX GREATER -1)
	add_subdirectory (uv)
endif ()
```
to 
```cmake
set (INCLUDED "")
generic_check_if_included (BINDINGS "io_uv") # with additional parameters
if (INCLUDED)
	add_subdirectory (uv)
endif ()
```

markus2330 replied:
Yes, this is certainly an improvement to status quo. Please make sure to not directly modify BINDINGS (because this is a user-specified CACHE variable we want to preserve exactly in the way like the user wrote it.) So running cmake again should find more bindings if dependencies were installed meanwhile.

Some remarks about your API "generic_check_if_included":

- In CMake you usually pass the OUTVARIABLE so that one can choose where the information is written to (and not use a hardcoded string like "INCLUDED".) plugin_check_if_included has the same problem, though.
- `generic_` can be omitted.
- It would be okay if "bindings" is a part of the function name. It is enough to reuse the implementation (so that plugin_check_if_included, binding_check_if_included, and tool_check_if_included is only very short calling the generic functions)  -- there no need to cram everything into one API.

markus2330 commented:
Build started sha1 is merged.

waht commented:
Okay, I added a small example program (`exampleio_uv` source is in src/bindings/io/example). It uses a file descriptor watch operation to read data from stdin and prints it using a timer. 

The build checks are failing due to the glib bindings (#1739).

Since we're now using plugin logic for bindings (`check_item_is_excluded`) `-DBINDINGS=ALL` now builds all bindings instead of only stable ones. To build stable bindings one needs to use `-DBINDINGS=STABLE`. I updated the docu where necessary.

@markus2330 please review if time

markus2330 commented on scripts/configure-debian:
```diff
@@ -12,7 +12,7 @@ cd $BUILD
 	-DPLUGINS="ALL" \
 	-DTOOLS="ALL" \
-	-DBINDINGS="ALL" \
+	-DBINDINGS="STABLE" \
```
Maybe ALL is not so bad here (even if it really means ALL).

For the default, we do not have to provide "-DBINDINGS" at all.

markus2330 commented on src/bindings/README.md:
```diff
@@ -9,7 +9,7 @@ in the respective programming languages, but you additionally need an
 can be immediately used in applications without plugins.
 
-List of currently supported bindings (included in `ALL`):
+List of currently supported bindings (included in `STABLE`):
```
included by default?

markus2330 commented on src/bindings/intercept/env/README.md:
```diff
@@ -1,5 +1,6 @@
 - infos = Information about elektrify getenv below
 - infos/author = Markus Raab <elektra@libelektra.org>
+- infos/status = stable intercept
```
Ideally we bring the other README.md also to this format?

waht replied:
Allright. I'm almost done finding out the authors.
I'm not sure about the author of the swig_python2 binding though. According to git log you created the files (e.g. CMakeLists.txt), but Manuel Mausz created the original python-legacy binding which got renamed to python2 right?

markus2330 replied:
Yes, Manuel Mausz (@manuelm) is the sole author of the python2 binding.

markus2330 commented on src/bindings/io/README.md:
```diff
@@ -0,0 +1,12 @@
+# I/O Bindings
```
What do you think about including this information to the README.md one level above?

Having it here would be possible, too. But then the README.md should be longer (to avoid the "too many tabs open" problem we already got complaints about) and somehow needs to be included into the website.

markus2330 commented on src/bindings/io/doc/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+## Build library
+
+file (GLOB IO_VARIANT_HDR_FILES *.h)
+file (GLOB IO_VARIANT_SRC_FILES *.c)
```
I am okay with GLOB (and we already use it often). I only wanted to mention that CMake guides discourage to do that because it won't track if you add new source files. (you'll need to invoke cmake manually.)

waht replied:
I see. I removed it from I/O bindings (no benefit with one source and one header file) but left it for gathering the test sources - when there are more I/O bindings we can add a module with functions for repetitive parts (e.g. gathering test sources).

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+#include <kdbio.h>
+#include <kdblogger.h>
+
+// Example I/O management library data structure
```
What about using Doxygen comments everywhere? Or are all these structs specific to the doc plugin?

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+static void ioDocBindingFdCallback (SomeIoLibHandle * handle, int bitmask)
+{
+	// For this example let's assume handle was passed as argument
+	assert (handle->data != NULL);
```
Please use ELEKTRA_ASSERT (also at other places)

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+	// You can store additional data for each operation in your operationData structure
+	operationData->operation.timer = timerOp;
+
+	// Here you need to add the operation to the I/O management library
```
These comments won't be part of Doxygen docu.

You can also include snippets of code (even of this file) using Doxygen's @snippet functionality.

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+ * @param  foo Some data from I/O management library (e.g. a handle)
+ * @return     Populated I/O interface
+ */
+ElektraIoInterface * elektraIoDocNew (char * foo)
```
Take a look at how ELEKTRA_PLUGIN_EXPORT works (src/include/kdbplugin.h). Even if we do not introduce the static/full variants for I/O bindings, code using ELEKTRA_IO_BINDING_EXPORT would be more compact and declarative.

waht replied:
I see. Varargs would make the code more declarative and compact but they sacrifice type safety. Using a struct (a struct just for this purpose) could be the middle ground, while the current solution is very compact but not declarative.

For plugins varargs provide the additional advantage of being able to omit optional functions, which is currently not needed. 
But if a new (and optional) function was necessary for I/O bindings we can maintain ABI compatibility. Maintaining ABI compatibility for structs seems [quite painful](https://stackoverflow.com/questions/25947997/how-does-sizeofstruct-help-provide-abi-compatibility) compared to the varargs solution. With functions ABI compatibility holds only as long as the prototype stays the same.

In summary I would trade type safety for possible ABI compatibility. What do you think?

markus2330 replied:
If you do not need optional parameters having everything as parameters sounds okay. Other differences of ELEKTRA_PLUGIN_EXPORT are:

1. allocation hidden within the function
2. the method name (elektraIoDocNew) is hidden by the macro. Doing so allows us to have static linkage and variants.

waht replied:
Okay, thank you.

Allocation is already done in `elektraIoNewBinding`. Other allocations are specific to bindings, but could be moved to an additional "open" function (like done with plugins).
Hiding the method name also comes with additional work when loading the binding (e.g. dl.c / static.c). 

For now there I see little reason for static linking or variantes. Since the interface is experimental we can still change this later.

markus2330 replied:
Ahh, I see. I was confused about that elektraIoDocNew also implements the "Open" of the binding.

If there is no reason to separate these concepts it can stay as it is.

waht replied:
Ok, great! :)

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+}
+
+/**
+ * Cleanup
```
Cleanup what was allocated before? Or more?

Please also use `@see` to link to related functions.

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+}
+
+/**
+ * Remove idle operation from IO-Binding
```
Unclear what this does? What is an idle operation?

markus2330 commented on src/bindings/io/doc/io_doc.c:
```diff
@@ -0,0 +1,365 @@
+}
+
+/**
+ * Calls the associated operation callback
```
A dot at the end might be needed so that Doxygen correctly knows what `@brief` is. Or are both lines in brief on purpose?

markus2330 commented on src/bindings/io/uv/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+
+if (${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
+	remove_binding(io_uv "cannot be built on Windows")
+elseif(NOT LIBUV_FOUND)
```
Spaces before "(". (Yes, some older CMake code does this still wrongly, fixes are welcomed.)

waht replied:
I'm not sure: Is the rule space or no space before "("?

markus2330 replied:
There should be a space before "(".

Rationale: We do not have detailed guidelines for CMake but in case of doubt we use the guidelines we already have for C.

markus2330 commented on src/bindings/io/uv/example/exampleio_uv.c:
```diff
@@ -0,0 +1,131 @@
+
+#define BUFFER_LENGTH 255
+#define ONE_SECOND 1000
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
```
Why implement as macro?

waht replied:
No reason at all - some result on stackoverflow suggested this macro... :)
I switched to a function.

markus2330 commented on src/bindings/io/uv/example/exampleio_uv.c:
```diff
@@ -0,0 +1,131 @@
+#include <unistd.h>
+
+#include <kdb.h>
+#include <kdbhelper.h>
```
Please document why these header files need to be included.

markus2330 commented on src/bindings/io/uv/example/exampleio_uv.c:
```diff
@@ -0,0 +1,131 @@
+#ifdef HAVE_LIBUV1
+	uv_loop_close (loop);
+#endif
+#ifdef HAVE_LIBUV0
```
elif? Or even else? (do they plan to change "uv_loop_close" again?)

markus2330 commented on src/include/kdbio.h:
```diff
@@ -0,0 +1,495 @@
+/**
+ * Check if file descriptor watch operation is enabled or disabled
+ * @param  fdOp    file descriptor operation
+ * @return         0 to disabled, any other value for enabled
```
retval should be used here.

markus2330 commented on src/include/kdbio.h:
```diff
@@ -0,0 +1,495 @@
+typedef ElektraIoInterface * (*ElektraIoTestSuiteCreateBinding) (void);
+
+/**
+ * Start io processing (e.g. event loop)
```
IO?

markus2330 commented on src/include/kdbio.h:
```diff
@@ -0,0 +1,495 @@
+ * @param start         Pointer to the start function
+ * @param stop          Pointer to the stop function
+ */
+void elektraIoTestSuite (ElektraIoTestSuiteCreateBinding createBinding, ElektraIoTestSuiteStart start, ElektraIoTestSuiteStop stop);
```
should be in a separate header file, or somewhere internally.

waht replied:
I inteded for it to be public so that application developers can test custom I/O binding implementations against the test suite. A different file would be better though (e.g kdbiotest.h)

markus2330 commented on src/libs/io/io.c:
```diff
@@ -0,0 +1,620 @@
+		return NULL;
+	}
+
+	ElektraIoInterface * binding = elektraMalloc (sizeof (*binding));
```
elektraCalloc already does the memset.

markus2330 commented on src/libs/io/io.c:
```diff
@@ -0,0 +1,620 @@
+		return NULL;
+	}
+
+	ElektraIoFdOperation * fdOp = elektraMalloc (sizeof (*fdOp));
```
again elektraCalloc (isn't memset even useless? You set all fields of the struct, don't you?)

waht replied:
Not all, but I now do and got rid of memset.

markus2330 commented on src/libs/io/io.c:
```diff
@@ -0,0 +1,620 @@
+	if (fdOp == NULL)
+	{
+		ELEKTRA_LOG_WARNING ("operation was NULL");
+		return 1;
```
why are error cases 1 here?

markus2330 commented on src/libs/io/io.c:
```diff
@@ -0,0 +1,620 @@
+		return 1;
+	}
+
+	fdOp->flags = flags;
```
What if flags is invalid? Do we know here which flags are valid?

markus2330 commented on src/plugins/constants/constants.c.in:
```diff
@@ -133,7 +133,7 @@ static void elektraAddConstants (KeySet * ks, Key * parentKey)
 	key = keyDup (parentKey);
 	keyAddName (key, "cmake/BINDINGS");
-	keySetString (key, "@BINDINGS@");
+	keySetString (key, "@ADDED_BINDINGS@");
```
Very careful work, thank you!

waht commented on cmake/Modules/LibAddBinding.cmake:
```diff
@@ -0,0 +1,243 @@
+# Example:
+#   add_binding_manual ("anynameyouwant")
+#
+function (add_binding_manual BINDING_NAME)
```
If the binding was added manually it should be removed with remove_binding. Ideally should be checked before using add_binding_manual.
A use case for this is the gsettings binding: If the user includes this binding the glib binding is automatically included (by adding the subdirectory and using add_binding_manual). In this case dependencies are checked after adding the binding.

waht commented on doc/COMPILE.md:
```diff
@@ -305,13 +305,10 @@ To specify specific tools you can use, e.g.:
 
-Bindings are used in the same way as `TOOLS`.
-For example, to include all non-experimental bindings you can use:
+Bindings are used in a similar way as `PLUGINS`.
```
The only difference would be that `infos/provides` is not checked/present for bindings. Does this make a difference from a user point of view?

markus2330 replied:
It might make sense to say e.g. `io_uv` provides `io`. It would definitely make sense to have semantics of PLUGINS and BINDINGS consistent.

waht replied:
Ok I see. I thought `info/status` would contain general category-like information, but the mix of statuses (from CONTRACT.ini) and categories (i.e. intercept and io) is not great. I will split those and add infos/provides just like for plugins.

waht commented on doc/COMPILE.md:
```diff
@@ -321,20 +318,26 @@ The SWIG executable may be specified with:
+Other available categories of bindings are:
+
+    - `DEFAULT` to build default bindings
+		- `STABLE` to build all stable bindings (without experimental and deprecated
```
Ah its not subcategory the identation is a formatting error I missed.

I don't mind changing it. I tried to replicate the previous behavior where possible:

- DEFAULT: cpp (was used when nothing was passed for BINDINGS)
- STABLE (former ALL): cpp, jna, glib, SWIG, INTERCEPT
- ALL: now every binding
- EXPERIMENTAL: gsettings, haskell, io_uv
- DEPRECATED: GI

"ALL;-EXPERIMENTAL" is good for consistency, but it includes more bindings than the previous default (and deprecated bindings).

markus2330 replied:
Thank you. STABLE is, however, not listed in `doc/CONTRACT.ini`? If it is only a synonym for `ALL;-EXPERIMENTAL;-DEPRECATED` we do not need it or did you have something else in mind?

jna is actually experimental, it was only an oversight in the last release. intercept_fs also is experimental (see at the end of README.md)

glib/SWIG is currently unmaintained (I will ask @manuelm about them).

So maybe we make the default to be `MAINTAINED;-EXPERIMENTAL;-DEPRECATED` (for both plugins and bindings). DEFAULT would not be needed then. Any opinions on that? (@sanssecours ?)

@waht If the discussion here gets to long, please move such proposals to a separate issue.

sanssecours replied:
> So maybe we make the default to be `MAINTAINED;-EXPERIMENTAL;-DEPRECATED` (for both plugins and bindings). 

Sounds good to me.

> DEFAULT would not be needed then. Any opinions on that?

I like the addition of `DEPRECATED`. I do not think we need to add `STABLE` or `DEFAULT`, since we can use `MAINTAINED;-EXPERIMENTAL;-DEPRECATED` and `cpp` instead.

waht replied:
> Thank you. STABLE is, however, not listed in `doc/CONTRACT.ini`? If it is only a synonym for `ALL;-EXPERIMENTAL;-DEPRECATED` we do not need it or did you have something else in mind?

I did not think of the statuses in `CONTRACTS.ini` but it makes sense to apply them to bindings as well. Yes, `STABLE` it would be just a synomym.
I'll update the usedBy in `CONTRACTS.ini`.

> I like the addition of DEPRECATED. I do not think we need to add `STABLE` or `DEFAULT`, since we can use `MAINTAINED;-EXPERIMENTAL;-DEPRECATED` and `cpp` instead.

Ok, great. I'll update the build scripts to have either no value (the new default `MAINTAINED;-EXPERIMENTAL;-DEPRECATED`) or where currently no value is present `cpp`. This will keep the build process reasonably quick.

> @waht If the discussion here gets to long, please move such proposals to a separate issue.

The BINDINGS overhaul got a bit more thorough than I expected :)

waht commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,214 @@
+A wrapper for I/O operations allows asynchronous notification processing by
+compatible plugins.
+The I/O-wrapper consists of an *interface* used by transport plugins and a
+*binding* which implements the actual I/O management functions for a specific
```
Sort of but I had I/O-bindings in mind - I've made the section more clear.

waht commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,214 @@
+
+//  ... initialization of KDB and I/O-binding
+
+Key * someKey = keyNew ("/sw/someprogram/v1/someKey", KEY_END);
```
Thanks! I changed the key's path to "/sw/myorg/myprogram/#1/current/somekey".

waht commented on doc/tutorials/notifications.md:
````diff
@@ -0,0 +1,214 @@
+```C
+void myCallback (Key * changedKey)
+{
+	// do something
````
I improved the example. I found a valid yet simple use case: terminal text color :)

waht commented:
Although there were just small things, there were many of them :)

- Added a Doxygen group for the doc binding (like doc.h for plugins)
- Improved overall Doxygen documentation in headers and c files
- Added kdbiotest.h with declarations for the test suite
- Added full infos/* fields to all binding's README.md files
- CMake: Use "MAINTAINED;-EXPERIMENTAL;-DEPRECATED" for as default for BINDINGS - build "ALL" in all "configure-*" scripts expcept for "configure-debian" where the default value is used.
- CMake: Switched from `try_add_binding` to `check_binding_included` and `add_binding` for more robustness
- CMake: Did some CMakeLists-formatting in src/bindings/

@markus2330 please review - especially the infos/status in README.md files.

sanssecours commented:
@waht 👋 Hi Thomas, looks like there are conflicts in `doc/docker/Dockerfile`. I think it would make sense to rebase your branch on `master`, before Markus reviews your changes.

waht commented:
Yep, sure. I've rebased the branch on `master`.

markus2330 commented on src/bindings/io/doc/elektra-io-doc.pc.in:
```diff
@@ -0,0 +1,14 @@
+prefix=@CMAKE_INSTALL_PREFIX@
```
Is this file needed?

waht replied:
No. At first I thought of it as a template for new bindings, but I removed unnecessary files. The binding is for documentation.

markus2330 commented on scripts/configure-home:
```diff
@@ -16,7 +16,7 @@ cmake	-DSWIG_EXECUTABLE=/usr/bin/swig3.0 \
 	-DKDB_DB_SPEC="~/.config/kdb/spec" \
 	-DTOOLS="ALL" \
-	-DBINDINGS="ALL" \
+	-DBINDINGS="STABLE" \
```
I thought we'll remove STABLE?

waht replied:
Ops, sure just a leftover.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -0,0 +1,229 @@
+yet.**
+
+This document explains how notifications are implemented in Elektra and how
+they can be used by application developers.
```
What is the difference between io lib and io binding?

waht replied:
`libelektra-io` contains utility functions for creating operations/bindings and accessing their fields (e.g. `elektraIoNewFdOperation`, `elektraIoNewBinding` and `elektraIoTimerIsEnabled`) while I/O bindings contain the functions (passed to `elektraIoNewBinding`) for a specific I/O management library.

markus2330 replied:
Thanks, that should be a part of the tutorial or docu.

markus2330 commented on doc/Doxyfile:
```diff
@@ -769,6 +769,7 @@ INPUT                  = "@PROJECT_SOURCE_DIR@/src/libs" \
                          "@PROJECT_SOURCE_DIR@/src/include" \
                          "@PROJECT_SOURCE_DIR@/src/bindings/cpp/include" \
                          "@PROJECT_SOURCE_DIR@/src/bindings/jna/libelektra4j/src" \
+                         "@PROJECT_SOURCE_DIR@/src/bindings/io/doc/io_doc.c" \
```
Did you consider to include the whole doc folder?

markus2330 commented on doc/AUTHORS.md:
```diff
@@ -207,3 +207,10 @@ ruby bindings, ruby plugin, build server, Puppet module
 - github user: [BernhardDenner](https://github.com/BernhardDenner)
 - devel/test on: Ubuntu and Debian
 
+## Thomas Wahringer
```
👍

markus2330 commented:
jenkins build all please

markus2330 commented:
I removed ready to merge because the changes in the scripts seem to fail some of the build jobs.

markus2330 commented:
When compiling and running everything locally I get no problems. So the only left-over problem seems to be that the build scripts include more bindings than they did before (which results to failed test cases).

Btw. this PR likely has conflicts with #1698, which one should we merge first?

waht commented:
I added the entries to the news for the next release. I added a highlight explaining the essentials but if it is too much you can still edit it before release.

@haskell PR: I looked at possible conflicting files and it does not seem too bad. Maybe there will be conflicts due to my reformatting. I don't mind resolving the conflicts later, but the longer we wait the more conflicts will arise.

Looking at the build logs the ruby bindings seem to fail in conjunction with ASAN:

- https://build.libelektra.org/jenkins/job/elektra-homepage/293/consoleFull
- https://build.libelektra.org/jenkins/job/elektra-gcc-configure-xdg/926/consoleFull
- https://build.libelektra.org/jenkins/job/elektra-clang-asan/351/consoleFull

@BernhardDenner can you have a look please?

markus2330 commented:
Can you create a separate ticket for the ruby bindings?

In this PR simply deactivate the bindings, as it was the case before.

>  I don't mind resolving the conflicts later, but the longer we wait the more conflicts will arise.

Yes, we should merge this as soon as possible.

waht commented:
Done!

- Added info about `elektra-io` library
- Exclude `swig_ruby` bindings until #1770 is resolved (with TODO in ElektraCache.cmake)

markus2330 commented:
Thank you, seems like the exclusion did not work? There is a failing test case saying: "
error: check of plugin ruby produced: "src/plugins/ruby/ruby.cpp:475:libelektra_Ruby_LTX_elektraPluginOpen: could not init ruby environment"
testscr_check_kdb_internal_check.sh RESULTS: 182 test(s) done 1 error(s)."

Please add the todos in doc/todo/TESTING

Please als run "jenkins run all please" then.

markus2330 commented:
jenkins build gcc-configure-debian-log please

waht commented:
Ok, I added the todo to TESTING (hope I did it right).

The build error is related to the ruby plugin. Unfortunately I could reproduce the problem on master and created #1771.

markus2330 commented:
Thank you for creating #1771. I hope it is an easy fix. The idea was to not merge PRs anymore which do not pass the build servers. On the other hand, the problem was there before. Maybe keeping ruby away from default plugins (e.g. by marking at as experimental) is the way to go. (If the trivial logging fix does not work.)

I start the rest of the builds to see if this problem is the only one.

markus2330 commented:
jenkins build all please

markus2330 commented:
jenkins build docker please

waht commented:
Sure this policy makes sense. The ruby plugin is already experimental; the build jobs just build all plugins.

waht commented:
jenkins build all please

sanssecours commented:
jenkins build multiconfig-gcc-stable please

markus2330 commented:
Great job!

After 16 hours finally all build jobs successfully completed 😄

waht commented:
Wohoop!
